### ISSUE TITLE
Missing sync events during historical batch imports

### ISSUE THREAD

**Issue by Fizzadar** [matrix-org/synapse#12281]
Split out from https://github.com/matrix-org/synapse/issues/9424#issuecomment-1061992367 onwards. I have inlined the relevant context from that issue below.

We have re-created the following issue; I've included the event list and trimmed sync responses in this gist: https://gist.github.com/Fizzadar/316dabfe9be30ebaa48e288de2380de1. Some notes:

+ The first sync containing the invite seems to correctly pull everything
+ The second sync includes member information but misses out the prior state
+ The invite also appears to arrive after the room creation event
+ Prior syncs do not reference the room at all (can provide JSON dumps if useful)

Notably, this is only occuring during initial backfill using MSC2716 batch sends. When this is disabled on the bridge we have not yet replicated the problem seen here.

The DAG:

```
synapse=> SELECT stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
LEFT JOIN event_json AS ej USING (event_id)
WHERE events.room_id ='!SjekxtmsBigrUMcMkX:beeper-dev.com' ORDER BY stream_ordering ASC;
 stream_ordering |             type             |                          state_key                           |                   event_id                   |                   prev_events
-----------------+------------------------------+--------------------------------------------------------------+----------------------------------------------+--------------------------------------------------
         -113203 | org.matrix.msc2716.insertion |                                                    | $G5GLwX6lc8MXKqwauoM1p9lkLD6S5GfwcosFlPra0ws | []
         -113202 | m.room.encrypted             |                                                    | $I3PsF44AMu5I_XOZQ2Hm1cOhcdLJCbA8NcAMS69hLEE | ["$G5GLwX6lc8MXKqwauoM1p9lkLD6S5GfwcosFlPra0ws"]
         -113201 | m.room.encrypted             |                                                    | $hIO0ZI-ZYr8gWGkWtvBiPfCA0jKdjiZP7PCnEfuutTY | ["$I3PsF44AMu5I_XOZQ2Hm1cOhcdLJCbA8NcAMS69hLEE"]
         -113200 | m.room.encrypted             |                                                    | $-jWSmA7NJ1dbl5b86K8FAKN2ydt-qoOw8ftCdSEN4A4 | ["$hIO0ZI-ZYr8gWGkWtvBiPfCA0jKdjiZP7PCnEfuutTY"]
         -113199 | m.room.encrypted             |                                                    | $5EtpKKW7g5HV5oCweU6vlbGjptR095p-kZSSSy2tBgs | ["$-jWSmA7NJ1dbl5b86K8FAKN2ydt-qoOw8ftCdSEN4A4"]
         -113198 | m.room.encrypted             |                                                    | $5aVzqGvybXzV3Xj3AwA8SQKqp2X2Z7RLH7IGX5VBeX4 | ["$5EtpKKW7g5HV5oCweU6vlbGjptR095p-kZSSSy2tBgs"]
         -113197 | org.matrix.msc2716.batch     |                                                    | $KkSRlRidPiOkkLEuiHs3gWLyCqCdQTHAVYJxzhDIzcM | ["$5aVzqGvybXzV3Xj3AwA8SQKqp2X2Z7RLH7IGX5VBeX4"]
         -111135 | org.matrix.msc2716.insertion |                                                    | $LQBf_c9-1stBXkPOLqZwkJgMg5YBCp448qcUwKLN6ks | []
         -111134 | m.room.encrypted             |                                                    | $RzKCB8rV0a9jfQVfbJeGvgGIZtT_X7ODb-78LUhXpXs | ["$LQBf_c9-1stBXkPOLqZwkJgMg5YBCp448qcUwKLN6ks"]
         -111133 | org.matrix.msc2716.batch     |                                                    | $l_jGmPdlA73wDzS3_gRwC67co1pTnx9-QBufta10xWI | ["$RzKCB8rV0a9jfQVfbJeGvgGIZtT_X7ODb-78LUhXpXs"]
         -111132 | org.matrix.msc2716.insertion |                                                    | $_mT6Awgb70IjwseWpCUeU0O4tm_iT7OBe4IPLJYUHUU | ["$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68"]
         2633404 | m.room.create                |                                                    | $ZMCw8gIYOeNmC2mGNKiAdxbRdwuhDiA9gNaM8EwDlRM | []
         2633421 | m.room.member                | @_adam4_whatsapp_xxx:beeper-dev.com                | $a6LZbYMvr5yYRcFc0uethSHLVC_0l7LRUr0nCZXB2lQ | ["$ZMCw8gIYOeNmC2mGNKiAdxbRdwuhDiA9gNaM8EwDlRM"]
         2633434 | m.room.power_levels          |                                                    | $S2rnt0m3dfbgC0ySGI3RGSdIjNqhgjjilZ2Gm7jGtvg | ["$a6LZbYMvr5yYRcFc0uethSHLVC_0l7LRUr0nCZXB2lQ"]
         2633448 | m.room.join_rules            |                                                    | $5hkAty1mFIMtuFINEh8wA3OyIgnO6Qxh1HgCTT9c1p0 | ["$S2rnt0m3dfbgC0ySGI3RGSdIjNqhgjjilZ2Gm7jGtvg"]
         2633461 | m.room.history_visibility    |                                                    | $nNR6iQSvWPpoKRvfpXJXU_sWpnk8eCwTFJD9zwC2rfo | ["$5hkAty1mFIMtuFINEh8wA3OyIgnO6Qxh1HgCTT9c1p0"]
         2633469 | m.room.guest_access          |                                                    | $iNJMDkjnFwBekoCZASqZmiNtH8UMLsPc_1VHJD6IFpU | ["$nNR6iQSvWPpoKRvfpXJXU_sWpnk8eCwTFJD9zwC2rfo"]
         2633478 | m.bridge                     | net.maunium.whatsapp://whatsapp/xxx@s.whatsapp.net | $p5fs4ik0Lnj-mIbw-cjPHX_aWzx0x6O_rIrhSi8_1C8 | ["$iNJMDkjnFwBekoCZASqZmiNtH8UMLsPc_1VHJD6IFpU"]
         2633492 | uk.half-shot.bridge          | net.maunium.whatsapp://whatsapp/xxx@s.whatsapp.net | $tAZ3y00piOU_54qs0MuK-HRPivXNRBmN8x6yY5KWXkk | ["$p5fs4ik0Lnj-mIbw-cjPHX_aWzx0x6O_rIrhSi8_1C8"]
         2633503 | m.room.avatar                |                                                    | $BZOaOEhozp28cCJtdh8A6R2hKgXiuDmd1NWI2vEuyPs | ["$tAZ3y00piOU_54qs0MuK-HRPivXNRBmN8x6yY5KWXkk"]
         2633515 | m.room.encryption            |                                                    | $qTv8IJZIdlH_3S7atbGTrd-7p7UEt4siocS0vz8-K3Y | ["$BZOaOEhozp28cCJtdh8A6R2hKgXiuDmd1NWI2vEuyPs"]
         2633523 | m.room.name                  |                                                    | $joClbFehqiqiVOuYYDWGJeYaozJnaccZgVog8j4bB6g | ["$qTv8IJZIdlH_3S7atbGTrd-7p7UEt4siocS0vz8-K3Y"]
         2633533 | m.room.topic                 |                                                    | $WP0ALeVvT2oCRSt9qs4_bE2O5AKsu_4GqbTCPFwaKa4 | ["$joClbFehqiqiVOuYYDWGJeYaozJnaccZgVog8j4bB6g"]
         2633547 | m.room.member                | @_adam4_whatsapp_bot:beeper-dev.com                | $hq4kZZPt4ItYdBF_I_AVITFHOZOZ2b_Av8ss3F06byg | ["$WP0ALeVvT2oCRSt9qs4_bE2O5AKsu_4GqbTCPFwaKa4"]
         2633571 | m.room.member                | @adam4:beeper-dev.com                              | $DyFoSr8CbRUyexbx6_yuZxgS4S7lAEHDYgldiRGnppc | ["$hq4kZZPt4ItYdBF_I_AVITFHOZOZ2b_Av8ss3F06byg"]
         2633582 | m.room.member                | @adam4:beeper-dev.com                              | $51gzIcw5d1fvbNLqYXdBwma669-H8DZlz0UzsavMBbU | ["$DyFoSr8CbRUyexbx6_yuZxgS4S7lAEHDYgldiRGnppc"]
         2633590 | m.room.member                | @_adam4_whatsapp_bot:beeper-dev.com                | $Sx9mswlxUJs1rRL4UyHARmY7V0h8_cH_c_ZDnwTn3cc | ["$51gzIcw5d1fvbNLqYXdBwma669-H8DZlz0UzsavMBbU"]
         2633601 | fi.mau.dummy.portal_created  |                                                    | $a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68 | ["$Sx9mswlxUJs1rRL4UyHARmY7V0h8_cH_c_ZDnwTn3cc"]
         2633614 | fi.mau.dummy.pre_backfill    |                                                    | $E97WApcmVmCJ6nt2WL3TnuxhyCLnPl1dt-t6p25ZLwY | ["$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68"]
         2633633 | fi.mau.dummy.backfill_end    |                                                    | $NILDi9SfVDgaW72alIEsc5ESzdJSoLip_c7klMAeuPo | ["$E97WApcmVmCJ6nt2WL3TnuxhyCLnPl1dt-t6p25ZLwY"]
(30 rows)
```

Note that this DAG looks similar to rooms where this problem did not occur.

**Comment by Fizzadar** [comment]
Based upon reviewing the MSC and room batch code the `prev_event` field _should_ be present for all insertion batches. I think this is possibly a separate issue.

For the sync itself it seems something is confusing the sync handler because none of the batch evens should ever come down via sync (noted as an issue on the MSC). In our live tests we do see batches come down with sync requests which I now believe to be incorrect.

**Comment by reivilibre** [comment]
> Have just checked and replaying the same sync request (account is untouched since the original issue) does include the missing m.bridge events as expected. It also includes the very first [MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716) event, the only one with any prev_event content ($_mT6Awgb70IjwseWpCUeU0O4tm_iT7OBe4IPLJYUHUU).

Very curious, so you're saying that replaying the sync request is giving you all the expected state events now?

Would you be happy to share a new trimmed extract of it?

**Comment by reivilibre** [comment]
I've been making scribbles on this graph of the DAG (→ = prev-events of, so newer events are generally at the top), thought attaching it here may be of use if others want to look at this too.

<details>
<summary>DAG</summary>
![2022_03_12281_dagger](https://user-images.githubusercontent.com/38398653/159970526-8fdafcdb-e0e0-43e2-95e3-b410ad62ab09.svg)
</details>

The scribbles themselves are probably less useful to others for the moment..

**Comment by Fizzadar** [comment]
> Very curious, so you're saying that replaying the sync request is giving you all the expected state events now?
> Would you be happy to share a new trimmed extract of it?

My apologies it _does_ miss the same events (phew!), so the issue is consistent. Must've copied the wrong since token somewhere. Here's the sync response when I re-create it today: https://gist.github.com/Fizzadar/316dabfe9be30ebaa48e288de2380de1#file-2022-03-24-trimmed-sync-s2633573_17_338_6732159_1082514_541479_274711_265584_1-json

Working on getting the event content too will update here.

**Comment by MadLittleMods** [comment]
@Fizzadar It's very hard to follow that gist and what's expected vs missing. Is is possible to more summarize that in terms of requests that happened.

1. Room create
2. `/batch_send`
3. `/batch_send`
4. `/sync`
5. `/sync?since=a`
    - Missing `$xxx` which is `m.room.member` `invite`
6. etc (with more detail)

Then I can try to create a [Complement](https://github.com/matrix-org/complement) test to reproduce and understand better.

**Comment by Fizzadar** [comment]
@MadLittleMods sure thing, describing the order below and I've also included the raw bridge logs specific for the room in question:

1. Client starts, `initalSync` happens before any bridge or backfill
2. Bridge starts, client signin/etc flow
3. Bridge starts backfill process - note this creates + backfills up to 20 rooms in parallel
4. Room create, initial state includes the `m.bridge` event as expected
5. `/sync?since=s2633508_17_338_6732159_1082514_541479_274711_265584_1`
  a. [this file in the gist](https://gist.github.com/Fizzadar/316dabfe9be30ebaa48e288de2380de1#file-trimmed_sync_3-s2633508_17_338_6732159_1082514_541479_274711_265584_1-json), contains all the room state as expected in the invite state dict
7. Bridge adds the user to the room via `m.room.member` state
8. First `/batch_send?prev_event_id=$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68`
9. `/sync?since=s2633573_17_338_6732159_1082514_541479_274711_265584_1`
  a. [this file in the gist](https://gist.github.com/Fizzadar/316dabfe9be30ebaa48e288de2380de1#file-trimmed_sync_4-s2633573_17_338_6732159_1082514_541479_274711_265584_1-json), this is where the `m.bridge` state event (`$p5fs4ik0Lnj-mIbw-cjPHX_aWzx0x6O_rIrhSi8_1C8` is missing + more)
  b. replaying this request today gives the same result
11. Second `/batch_send?batch_id=IepWImCQ&prev_event_id=$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68`

Note that this is not consistent - of 20 rooms being created + backfilled in parallel there may be one or two that have missing state as described above.

I've added the [raw logs from the bridge specific to the affected room in the gist](https://gist.github.com/Fizzadar/316dabfe9be30ebaa48e288de2380de1#file-z_bridge_log-log).

**Comment by bradtgmurray** [comment]
I'm curious if the issue is using a sync token at a particular point in the event graph, and the inconsistency of whether we can reproduce it or not is just the frequency of the client syncing and where the graph for a particular room gets divided between syncs. I wonder if we could take a room that "worked" for us and then manually bump back the sync token up the event graph until we get a sync that doesn't include the m.bridge event anymore and should.

**Comment by MadLittleMods** [comment]
> I'm curious if the issue is using a sync token at a particular point in the event graph, and the inconsistency of whether we can reproduce it or not is just the frequency of the client syncing and where the graph for a particular room gets divided between syncs. I wonder if we could take a room that "worked" for us and then manually bump back the sync token up the event graph until we get a sync that doesn't include the m.bridge event anymore and should.

This sounds pretty plausible. `rooms.join.<roomId>.state.events` should only include:

> Updates to the state, **between the time indicated by the since parameter, and the start of the timeline** (or all state up to the start of the timeline, if since is not given, or full_state is true).
>
> *-- https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3sync*

When I look at the Synapse database dump of events, I'm not quite seeing that behavior. I've included and sorted in the stream tokens from `(sync 1 ...)` and `(sync 2 ...)` into the list. And I've added `(1)` and `(2)` if they appeared in the `/sync` `state` list accordingly.

```
synapse=> SELECT stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
LEFT JOIN event_json AS ej USING (event_id)
WHERE events.room_id ='!SjekxtmsBigrUMcMkX:beeper-dev.com' ORDER BY stream_ordering ASC;
 stream_ordering |             type             |                          state_key                           |                   event_id                   |                   prev_events
-----------------+------------------------------+--------------------------------------------------------------+----------------------------------------------+--------------------------------------------------
         -113203 | org.matrix.msc2716.insertion |                                                    | $G5GLwX6lc8MXKqwauoM1p9lkLD6S5GfwcosFlPra0ws | []
         -113202 | m.room.encrypted             |                                                    | $I3PsF44AMu5I_XOZQ2Hm1cOhcdLJCbA8NcAMS69hLEE | ["$G5GLwX6lc8MXKqwauoM1p9lkLD6S5GfwcosFlPra0ws"]
         -113201 | m.room.encrypted             |                                                    | $hIO0ZI-ZYr8gWGkWtvBiPfCA0jKdjiZP7PCnEfuutTY | ["$I3PsF44AMu5I_XOZQ2Hm1cOhcdLJCbA8NcAMS69hLEE"]
         -113200 | m.room.encrypted             |                                                    | $-jWSmA7NJ1dbl5b86K8FAKN2ydt-qoOw8ftCdSEN4A4 | ["$hIO0ZI-ZYr8gWGkWtvBiPfCA0jKdjiZP7PCnEfuutTY"]
         -113199 | m.room.encrypted             |                                                    | $5EtpKKW7g5HV5oCweU6vlbGjptR095p-kZSSSy2tBgs | ["$-jWSmA7NJ1dbl5b86K8FAKN2ydt-qoOw8ftCdSEN4A4"]
         -113198 | m.room.encrypted             |                                                    | $5aVzqGvybXzV3Xj3AwA8SQKqp2X2Z7RLH7IGX5VBeX4 | ["$5EtpKKW7g5HV5oCweU6vlbGjptR095p-kZSSSy2tBgs"]
         -113197 | org.matrix.msc2716.batch     |                                                    | $KkSRlRidPiOkkLEuiHs3gWLyCqCdQTHAVYJxzhDIzcM | ["$5aVzqGvybXzV3Xj3AwA8SQKqp2X2Z7RLH7IGX5VBeX4"]
         -111135 | org.matrix.msc2716.insertion |                                                    | $LQBf_c9-1stBXkPOLqZwkJgMg5YBCp448qcUwKLN6ks | []
         -111134 | m.room.encrypted             |                                                    | $RzKCB8rV0a9jfQVfbJeGvgGIZtT_X7ODb-78LUhXpXs | ["$LQBf_c9-1stBXkPOLqZwkJgMg5YBCp448qcUwKLN6ks"]
         -111133 | org.matrix.msc2716.batch     |                                                    | $l_jGmPdlA73wDzS3_gRwC67co1pTnx9-QBufta10xWI | ["$RzKCB8rV0a9jfQVfbJeGvgGIZtT_X7ODb-78LUhXpXs"]
         -111132 | org.matrix.msc2716.insertion |                                                    | $_mT6Awgb70IjwseWpCUeU0O4tm_iT7OBe4IPLJYUHUU | ["$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68"]
         2633404 | m.room.create (1)            |                                                    | $ZMCw8gIYOeNmC2mGNKiAdxbRdwuhDiA9gNaM8EwDlRM | []
         2633421 | m.room.member (1) (2)        | @_adam4_whatsapp_xxx:beeper-dev.com                | $a6LZbYMvr5yYRcFc0uethSHLVC_0l7LRUr0nCZXB2lQ | ["$ZMCw8gIYOeNmC2mGNKiAdxbRdwuhDiA9gNaM8EwDlRM"]
         2633434 | m.room.power_levels          |                                                    | $S2rnt0m3dfbgC0ySGI3RGSdIjNqhgjjilZ2Gm7jGtvg | ["$a6LZbYMvr5yYRcFc0uethSHLVC_0l7LRUr0nCZXB2lQ"]
         2633448 | m.room.join_rules (1)        |                                                    | $5hkAty1mFIMtuFINEh8wA3OyIgnO6Qxh1HgCTT9c1p0 | ["$S2rnt0m3dfbgC0ySGI3RGSdIjNqhgjjilZ2Gm7jGtvg"]
         2633461 | m.room.history_visibility    |                                                    | $nNR6iQSvWPpoKRvfpXJXU_sWpnk8eCwTFJD9zwC2rfo | ["$5hkAty1mFIMtuFINEh8wA3OyIgnO6Qxh1HgCTT9c1p0"]
         2633469 | m.room.guest_access          |                                                    | $iNJMDkjnFwBekoCZASqZmiNtH8UMLsPc_1VHJD6IFpU | ["$nNR6iQSvWPpoKRvfpXJXU_sWpnk8eCwTFJD9zwC2rfo"]
         2633478 | m.bridge (1)                 | net.maunium.whatsapp://whatsapp/xxx@s.whatsapp.net | $p5fs4ik0Lnj-mIbw-cjPHX_aWzx0x6O_rIrhSi8_1C8 | ["$iNJMDkjnFwBekoCZASqZmiNtH8UMLsPc_1VHJD6IFpU"]
         2633492 | uk.half-shot.bridge (1)      | net.maunium.whatsapp://whatsapp/xxx@s.whatsapp.net | $tAZ3y00piOU_54qs0MuK-HRPivXNRBmN8x6yY5KWXkk | ["$p5fs4ik0Lnj-mIbw-cjPHX_aWzx0x6O_rIrhSi8_1C8"]
         2633503 | m.room.avatar (1)            |                                                    | $BZOaOEhozp28cCJtdh8A6R2hKgXiuDmd1NWI2vEuyPs | ["$tAZ3y00piOU_54qs0MuK-HRPivXNRBmN8x6yY5KWXkk"]
-------- 2633508 (sync 1 while invited)
         2633515 | m.room.encryption (1)        |                                                    | $qTv8IJZIdlH_3S7atbGTrd-7p7UEt4siocS0vz8-K3Y | ["$BZOaOEhozp28cCJtdh8A6R2hKgXiuDmd1NWI2vEuyPs"]
         2633523 | m.room.name (1)              |                                                    | $joClbFehqiqiVOuYYDWGJeYaozJnaccZgVog8j4bB6g | ["$qTv8IJZIdlH_3S7atbGTrd-7p7UEt4siocS0vz8-K3Y"]
         2633533 | m.room.topic (1)             |                                                    | $WP0ALeVvT2oCRSt9qs4_bE2O5AKsu_4GqbTCPFwaKa4 | ["$joClbFehqiqiVOuYYDWGJeYaozJnaccZgVog8j4bB6g"]
         2633547 | m.room.member (2)            | @_adam4_whatsapp_bot:beeper-dev.com                | $hq4kZZPt4ItYdBF_I_AVITFHOZOZ2b_Av8ss3F06byg | ["$WP0ALeVvT2oCRSt9qs4_bE2O5AKsu_4GqbTCPFwaKa4"]
         2633571 | m.room.member (1)            | @adam4:beeper-dev.com                              | $DyFoSr8CbRUyexbx6_yuZxgS4S7lAEHDYgldiRGnppc | ["$hq4kZZPt4ItYdBF_I_AVITFHOZOZ2b_Av8ss3F06byg"]
-------- 2633573 (sync 2 after being joined)
         2633582 | m.room.member (2)            | @adam4:beeper-dev.com                              | $51gzIcw5d1fvbNLqYXdBwma669-H8DZlz0UzsavMBbU | ["$DyFoSr8CbRUyexbx6_yuZxgS4S7lAEHDYgldiRGnppc"]
         2633590 | m.room.member                | @_adam4_whatsapp_bot:beeper-dev.com                | $Sx9mswlxUJs1rRL4UyHARmY7V0h8_cH_c_ZDnwTn3cc | ["$51gzIcw5d1fvbNLqYXdBwma669-H8DZlz0UzsavMBbU"]
         2633601 | fi.mau.dummy.portal_created  |                                                    | $a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68 | ["$Sx9mswlxUJs1rRL4UyHARmY7V0h8_cH_c_ZDnwTn3cc"]
         2633614 | fi.mau.dummy.pre_backfill    |                                                    | $E97WApcmVmCJ6nt2WL3TnuxhyCLnPl1dt-t6p25ZLwY | ["$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68"]
         2633633 | fi.mau.dummy.backfill_end    |                                                    | $NILDi9SfVDgaW72alIEsc5ESzdJSoLip_c7klMAeuPo | ["$E97WApcmVmCJ6nt2WL3TnuxhyCLnPl1dt-t6p25ZLwY"]
(30 rows)
```



### Actual reproduction gist questions

I'm still fuzzy on some details:

 1. Who is the creator of the room?
    - I assume `@_adam4_whatsapp_xxx:beeper-dev.com` (is this the same as `@_adam4_whatsapp_bot:beeper-dev.com`?)
 1. Who is the bot?
    - I assume `@_adam4_whatsapp_bot:beeper-dev.com`
 1. Who is doing the `/sync`'ing?
    - I assume `@adam4:beeper-dev.com`
 1. Are we using the `invite` field with `/createRoom`?
 1. In the `/sync` responses, are we expecting the `m.bridge` event to be in `rooms.join.<roomId>.timeline.events` or `rooms.join.<roomId>.state.events`?
 
 
 ### Complement test
 
 I wrote a Complement test to try to reproduce the problem but I am seeing some differences:
 
  1. Does the test setup look correct to what you're doing? https://github.com/matrix-org/complement/pull/352
  1. With the current test, in the first incremental sync, I'm not seeing the `m.bridge` be part of the first `rooms.invite.<roomId>.invite_state.events` like you do.
     - Is this necessary?
     - The spec isn't clear to me what should be included in `invite_state`
  1. In the second incremental sync,
     - `m.bridge` is part of `rooms.join.<roomId>.timeline.events`
     - The only event in `rooms.join.<roomId>.state.events` is the `m.room.create` event which makes sense because it shows the state at the start of the current `timeline` shown in the `/sync`.

**Comment by Fizzadar** [comment]
>     1. Who is the creator of the room?
>        * I assume `@_adam4_whatsapp_xxx:beeper-dev.com` (is this the same as `@_adam4_whatsapp_bot:beeper-dev.com`?)
>     2. Who is the bot?
>        * I assume `@_adam4_whatsapp_bot:beeper-dev.com`
>     3. Who is doing the `/sync`'ing? 
>        * I assume `@adam4:beeper-dev.com`

All correct.

>     4. Are we using the `invite` field with `/createRoom`?

Yes - for the bot user, the sync user `@adam4:beeper-dev.com` is invited after the room is created.

>     5. In the `/sync` responses, are we expecting the `m.bridge` event to be in `rooms.join.<roomId>.timeline.events` or `rooms.join.<roomId>.state.events`?

I expect to see the `m.bridge` state within the `rooms.join.<roomid>.state.events` - this is the behavior we see elsewhere when not using MSC2716.

_However_ this actually appears to be ad-odds with the spec. According to the spec I would not expect to see either the `m.bridge` event nor `$a6LZbYMvr5yYRcFc0uethSHLVC_0l7LRUr0nCZXB2lQ` (which is included). I'm unsure how, if the spec is followed exactly, an invited room member is ever able to receive any room state prior to their join (assuming one should ignore `invite_state`). Maybe I'm misinterpreting something here?

> I wrote a Complement test to try to reproduce the problem but I am seeing some differences:
> 
>     1. Does the test setup look correct to what you're doing? [Test case to try to reproduce missing state in sync (possibly related to MSC2716) complement#352](https://github.com/matrix-org/complement/pull/352)

This looks correct - I think the only thing missing is the `state_key` for the `m.bridge` event which would explain why it never appears in the sync state I  think.

>        * The spec isn't clear to me what should be included in `invite_state`

Agreed, I think it's just for display purposes but it's not clear. What is clear is any `invite_state` should not replace `rooms.join.<roomid>.state.events`.

**Comment by MadLittleMods** [comment]
@Fizzadar Just to make sure we're not chasing after an X/Y problem, what is the actual root problem that you are dealing with? Why does it matter that `m.bridge` is missing from the `/sync` `state` response?

@Fizzadar Can you post the Synapse db dump and the two /sync results when it's working correctly? It would be good to compare what's completely expected for you.

---

> The spec isn't clear to me what should be included in `invite_state`

Diving into this more on Synapse, it looks like `invite_state` depends on what you have [`room_prejoin_state`](https://github.com/matrix-org/synapse/blob/3c41d87b67d3a62edfc660b4fe8f2545f5dbee4f/docs/sample_config.yaml#L1520-L1546) (previously `room_invite_state_types`) configured as in `homeserver.yaml`.  Can you share whether you have something configured here?



> What is clear is any `invite_state` should not replace `rooms.join.<roomid>.state.events`.

👍 For my own reference, here are the API docs that explain that:

> `invite_state`
>
> These events do not replace any state that the client already has for the room, for example if the client has archived the room. Instead the client should keep two separate copies of the state: the one from the `invite_state` and one from the archived `state`. If the client joins the room then the current state will be given as a delta against the archived `state` not the `invite_state`.
>
> *-- https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3sync*


---

> I think the only thing missing is the `state_key` for the `m.bridge` event which would explain why it never appears in the sync state I think.

I think it's fine with `"state_key": "",` which stores it as the sorta root global one for the room. It's unclear to me why you define a `"state_key": "net.maunium.whatsapp://whatsapp/xxx@s.whatsapp.net"` in your case? Why should it be namespaced there?

We do the same `"state_key": "",` with [bridged Gitter rooms for example](https://gitlab.com/gitterHQ/webapp/-/blob/ca9d6e8039925e57ec8b7cb0a167237602585849/modules/matrix-bridge/lib/matrix-utils.js#L381-396).

**Comment by Fizzadar** [comment]
> @Fizzadar Just to make sure we're not chasing after an X/Y problem, what is the actual root problem that you are dealing with? Why does it matter that `m.bridge` is missing from the `/sync` `state` response?

We're using this to display the correct bridge icon in the UI which is how we discovered the issue.

> We do the same `"state_key": "",` with [bridged Gitter rooms for example](https://gitlab.com/gitterHQ/webapp/-/blob/ca9d6e8039925e57ec8b7cb0a167237602585849/modules/matrix-bridge/lib/matrix-utils.js#L381-396).

Yep sorry my bad, [the check is indeed against `None` explicitly](https://github.com/matrix-org/synapse/blob/develop/synapse/events/__init__.py#L336).

> @Fizzadar Can you post the Synapse db dump and the two `/sync` results when it's working correctly? It would be good to compare what's completely expected for you.

Absolutely, I've added the events and two syncs of a room that worked during the same bridge setup [in this gist](https://gist.github.com/Fizzadar/25f7f033f500f07bef4a548f91fc61e1). I have also added the annotations as you did above with the sync positions and highlighting where the state came down.

This is where things get interesting - the second sync contains only events under `timeline.events` - including events prior to the since token! In fact it looks like every room event is sent down in the timeline.

**Comment by bradtgmurray** [comment]
This feels like a hole in the spec, no? If you're joining a room for the first time, what's the expected behaviour of clients with respect to fetching the room state? It appears that clients are just assuming they'll get it based as part of sync (this is what we're seeing with our matrix-js-sdk fork) but that doesn't seem to be specified.

**Comment by Fizzadar** [comment]
I think I've just identified the situation this is occurring - it appears to happen in rooms where the first joined sync is not limited (ie `timeline.limited=False`). Looking at every room/sync we have a record of in our testing this is holding true. Digging into the codebase, the two paths (limited vs. not) differ greatly:

https://github.com/matrix-org/synapse/blob/8fe930c215f69913fbcd96d609ec6950644e4ec4/synapse/handlers/sync.py#L946-L1001

vs:

https://github.com/matrix-org/synapse/blob/8fe930c215f69913fbcd96d609ec6950644e4ec4/synapse/handlers/sync.py#L1002-L1021

In a non-limited/no-gap sync, only relevant membership state events are fetched. This matches the problem we are seeing here and indeed aligns with the spec (right or wrong). Whereas a limited response fetches a lot more state events, I'm currently trying to understand this codepath better so have no immediate conclusion on why it doesn't match the spec.

**Comment by Fizzadar** [comment]
Note: when I replay the working gist example above it no longer works (no `m.bridge` state event included). Dropping the sync token back one by one, as soon as both the invite + join events are ahead, things work as expected.

In fact - this appears to be the case for _every room_ that has been backfilled using `batch_send`. If I replay the sync using a token between invite and join events, the state is consistently missing. Given that most of the sync requests do not show this problem in our live client<>bridge setups (on which the gists are based), this must be related to the live ongoing batch sending at the time, which explains why this only happens to some rooms.

---

Had some time for some in-depth investigations; I've bootstrapped a local sync worker pointing at our development database used in all these examples and tracked down the problem to the `full_state` flag (previous comment), specifically where `newly_joined` is set. The following block is executed:

https://github.com/matrix-org/synapse/blob/8fe930c215f69913fbcd96d609ec6950644e4ec4/synapse/handlers/sync.py#L1764-L1774

In the failing sync from the original message in this issue (`since=s2633573_17_338_6732159_1082514_541479_274711_265584_1`) the `old_mem_ev_id` is the same as the latest membership event for the room (`$51gzIcw5d1fvbNLqYXdBwma669-H8DZlz0UzsavMBbU`). Thus the check on L1773 does not match when it should.

This situation seems to occur when a sync starts inbetween the invite event and join event. I can reliably replicate this for other rooms by rolling back the sync token to between the two events. The problem appears to be that `get_state_at(since_token)` is returning state that exists after the token. Following that leads to this:

https://github.com/matrix-org/synapse/blob/8fe930c215f69913fbcd96d609ec6950644e4ec4/synapse/handlers/sync.py#L686-L692

This is where MSC2716 possibly comes in because this returns the first `org.matrix.msc2716.insertion` event, `_mT6Awgb70IjwseWpCUeU0O4tm_iT7OBe4IPLJYUHUU`, and the `get_state_after_event` call based on that gets back the join event `$51gzIcw5d1fvbNLqYXdBwma669-H8DZlz0UzsavMBbU`.

I made [a small change here to handle this](https://github.com/Fizzadar/synapse/commit/a9a6ed945dfe02f2a583a2a4410597d83ff7c4e2) and the issue appears to be resolved in all my testing.

**Comment by MadLittleMods** [comment]
Great find @Fizzadar! Here are my own notes for understanding the problem as well:

## Why is Synapse not sending all of the initial state when `/sync?since=s123` between an `invite` and `join` event?


[`get_state_at(since_token)`](https://github.com/matrix-org/synapse/blob/287a9c1e20e1e271d69be581d6bb8ee5a8de9d88/synapse/handlers/sync.py#L673-L679) -> [`get_recent_events_for_room`](https://github.com/matrix-org/synapse/blob/287a9c1e20e1e271d69be581d6bb8ee5a8de9d88/synapse/storage/databases/main/stream.py#L654-L657) -> [`_paginate_room_events_txn`](https://github.com/matrix-org/synapse/blob/287a9c1e20e1e271d69be581d6bb8ee5a8de9d88/synapse/storage/databases/main/stream.py#L1095) will return events with the most recent events by topological ordering (same as `depth` in Synapse) (`DESC`, highest to lowest), then tie-break with `stream_ordering` (`DESC`, highest to lowest). The events are also filtered to be below the given `stream_ordering` determined by the `since` token.

The SQL ends up looking like:
```sql
SELECT
    event.event_id, event.instance_name,
    event.topological_ordering, event.stream_ordering
FROM events AS event

WHERE event.outlier = FALSE AND event.room_id = ? AND (2633573 >= event.stream_ordering)
ORDER BY event.topological_ordering DESC,
event.stream_ordering DESC LIMIT 2
```

For the problem `/sync` missing state, at this point in time while backfilling, `fi.mau.dummy.portal_created` (`$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68`) is the most recent event by `depth` in the room but it's excluded because of its `stream_ordering` (`2633573 >= 2633601` -> `False`). However, the `/batch_send` history is inserted next to `fi.mau.dummy.portal_created` via `/batch_send?prev_event_id=$a6xyChJTR8wJLsfYfGI8Tu1W5cXrtohkBRdwCSsyG68` which will inherit the same `depth` as that event but all of them will have a negative `stream_ordering` so they are included in the results as the latest in the room below the given `since` `stream_ordering` (`2633573 >= negative any number` -> `True`). You can see why the events are returned this way if we sort our database dump the same way `get_state_at(since_token)` is sorting:

```sh
$ psql synapse

synapse=> SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
LEFT JOIN event_json AS ej USING (event_id)
WHERE events.room_id ='!SjekxtmsBigrUMcMkX:beeper-dev.com'
ORDER BY depth DESC, stream_ordering DESC;

...
```


And since historical messages will resolve all of the state from where they were inserted at plus whatever was set in `/batch_send` `state_events_at_start`, it will include the `m.room.member` `join` event for the syncing user which the [`old_mem_ev` logic](https://github.com/matrix-org/synapse/blob/8fe930c215f69913fbcd96d609ec6950644e4ec4/synapse/handlers/sync.py#L1764-L1774) is getting confused about.


The problem will occur for anyone who joins the room before the `/batch_send` and hasn't synced before that to get the initial state.



### Is this unique to MSC2716?

This problem seems like it wouldn't be unique to using MSC2716 though. It seems like it could happen with federation backfilled messages as well since any backfilled message over federation also gets a negative `stream_ordering` (note that this is federation backfill which is different from the "backfill" you're doing to fill out the room history with MSC2716).

And is incremental sync is already know to have problems in this area: https://github.com/matrix-org/synapse/issues/3305



## Normal room where it works ✅



### Reproduction


<details>
<summary>Reproduction steps</summary>

 1. Add an application service to your Synapse instance
 1. With the application service `as_token`, create a room and invite a normal user
    `POST http://localhost:8008/_matrix/client/v3/createRoom`
    ```json
    {
        "preset":       "public_chat",
        "name":         "test-synapse-12281-v0.0.1",
        "room_version": "9",
        "invite": ["@ericgittertester:my.synapse.server"],
        "initial_state": [
            {
                "type":      "m.other.state",
                "state_key": "",
                "content": {
                    "foo": "bar"
                }
            }
        ]
    }
    ```
 1. In another room, send some messages to advance the `stream_ordering` in your Synapse instance. This will add some space for us to make stream tokens in between and use with `/sync?since=abc`
 1. Join the room as the normal user
 1. Dump the database, to easily see the gap in `stream_ordering` between the `m.room.member` `invite` and `join` events.
    ```sh
    $ psql synapse

    SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
    LEFT JOIN event_json AS ej USING (event_id)
    WHERE events.room_id ='!VpJGWzRaYnBJHwCaWd:my.synapse.server' ORDER BY stream_ordering ASC;
    ```
 1. As the normal user, sync in the problem gap spot, `GET http://localhost:8008/_matrix/client/r0/sync?filter=0&timeout=30000&since=s877_205797_22_82_396_8_19_160_1`
 1. Notice that `rooms.join.<roomId>.timeline.events` includes all initial state, including the `m.other.state` from the `initial_state` during room creation. 

</details>


### Notes

Here is an overview of the events in the room:

```sh
$ psql synapse

SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
LEFT JOIN event_json AS ej USING (event_id)
WHERE events.room_id ='!VpJGWzRaYnBJHwCaWd:my.synapse.server' ORDER BY stream_ordering ASC;


depth | stream_ordering |           type            |              state_key              |                   event_id                   |                   prev_events
-------+-----------------+---------------------------+-------------------------------------+----------------------------------------------+--------------------------------------------------
     1 |             863 | m.room.create             |                                     | $i0U5s25xbRoLLqancRLFOUcVEaLwNjZSzvvT5bU4_sw | []
     2 |             864 | m.room.member             | @erictroupetester:my.synapse.server | $RQXO-19yjf_zprhDkN4CFjj76zNyLYVwHSJRgGAarrE | ["$i0U5s25xbRoLLqancRLFOUcVEaLwNjZSzvvT5bU4_sw"]
     3 |             865 | m.room.power_levels       |                                     | $vxSLpdP0dYGLfY4SF5ajpzm4HIGBSl8ShXFngP6dwNc | ["$RQXO-19yjf_zprhDkN4CFjj76zNyLYVwHSJRgGAarrE"]
     4 |             866 | m.room.join_rules         |                                     | $Nzph2kpEaIG2FBl-1_N7Yu7Wq7q-0pLG1o-sjPOqYL0 | ["$vxSLpdP0dYGLfY4SF5ajpzm4HIGBSl8ShXFngP6dwNc"]
     5 |             867 | m.room.history_visibility |                                     | $4Hqnlf9LuYYN28G6yjFz50WKJEDI8m5vOCirNAK4vcU | ["$Nzph2kpEaIG2FBl-1_N7Yu7Wq7q-0pLG1o-sjPOqYL0"]
     6 |             868 | m.bridge                  |                                     | $EOa43fMXoJHizHb_s7VwLhRdLs6FMqjIxpGMhDC7tjo | ["$4Hqnlf9LuYYN28G6yjFz50WKJEDI8m5vOCirNAK4vcU"]
     7 |             869 | m.other.state             |                                     | $6T8IYDLBdLGfYcrHW3Ntj9XL1WCJAN7u670hNgDhz2A | ["$EOa43fMXoJHizHb_s7VwLhRdLs6FMqjIxpGMhDC7tjo"]
     8 |             870 | m.room.name               |                                     | $YCc5eyjIjx_qNQPtsnklJ7prZmhRluimQ246ytO9B6s | ["$6T8IYDLBdLGfYcrHW3Ntj9XL1WCJAN7u670hNgDhz2A"]
     9 |             871 | m.room.member             | @ericgittertester:my.synapse.server | $us9V0eqQbu_AWtxyquIDn-QFVdRL0KBnZE_AnXjefSw | ["$YCc5eyjIjx_qNQPtsnklJ7prZmhRluimQ246ytO9B6s"]
    10 |             880 | m.room.member             | @ericgittertester:my.synapse.server | $Usp1bf1j4SLhZLQ1IvGFhr2o-CyBjtISkBTOEnyf7z8 | ["$us9V0eqQbu_AWtxyquIDn-QFVdRL0KBnZE_AnXjefSw"]
    11 |             889 | m.room.message            |                                     | $Pzm2WJDUBhja2cwjEXR_suCAsd_SsV8pGdIQZ1ljXQQ | ["$Usp1bf1j4SLhZLQ1IvGFhr2o-CyBjtISkBTOEnyf7z8"]
(11 rows)
```

✅  `/sync?since=s877_xxx` (anything between the invite `871` and the join `880`)

Using the `get_state_at(since_token)` sort ( `ORDER BY depth DESC, stream_ordering DESC;`), you can see that the `m.room.member` `invite` is the most recent event in the room but will resolve just the `invite` state since the `join` happens later and is exlcuded because it it's `stream_ordering` is too high (see the filter note below).

 - The `-` red messages are filtered out because of the `WHERE 877 >= event.stream_ordering` in the SQL.
 - The green `+` is the selected event and is correct as expected

```diff
  $ psql synapse
  
  SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
  LEFT JOIN event_json AS ej USING (event_id)
  WHERE events.room_id ='!VpJGWzRaYnBJHwCaWd:my.synapse.server'
  ORDER BY depth DESC, stream_ordering DESC;
  
  
  depth | stream_ordering |           type            |              state_key              |                   event_id                   |                   prev_events
  -------+-----------------+---------------------------+-------------------------------------+----------------------------------------------+--------------------------------------------------
-     11 |             889 | m.room.message            |                                     | $Pzm2WJDUBhja2cwjEXR_suCAsd_SsV8pGdIQZ1ljXQQ | ["$Usp1bf1j4SLhZLQ1IvGFhr2o-CyBjtISkBTOEnyf7z8"]
-     10 |             880 | m.room.member             | @ericgittertester:my.synapse.server | $Usp1bf1j4SLhZLQ1IvGFhr2o-CyBjtISkBTOEnyf7z8 | ["$us9V0eqQbu_AWtxyquIDn-QFVdRL0KBnZE_AnXjefSw"]
+      9 |             871 | m.room.member             | @ericgittertester:my.synapse.server | $us9V0eqQbu_AWtxyquIDn-QFVdRL0KBnZE_AnXjefSw | ["$YCc5eyjIjx_qNQPtsnklJ7prZmhRluimQ246ytO9B6s"]
       8 |             870 | m.room.name               |                                     | $YCc5eyjIjx_qNQPtsnklJ7prZmhRluimQ246ytO9B6s | ["$6T8IYDLBdLGfYcrHW3Ntj9XL1WCJAN7u670hNgDhz2A"]
       7 |             869 | m.other.state             |                                     | $6T8IYDLBdLGfYcrHW3Ntj9XL1WCJAN7u670hNgDhz2A | ["$EOa43fMXoJHizHb_s7VwLhRdLs6FMqjIxpGMhDC7tjo"]
       6 |             868 | m.bridge                  |                                     | $EOa43fMXoJHizHb_s7VwLhRdLs6FMqjIxpGMhDC7tjo | ["$4Hqnlf9LuYYN28G6yjFz50WKJEDI8m5vOCirNAK4vcU"]
       5 |             867 | m.room.history_visibility |                                     | $4Hqnlf9LuYYN28G6yjFz50WKJEDI8m5vOCirNAK4vcU | ["$Nzph2kpEaIG2FBl-1_N7Yu7Wq7q-0pLG1o-sjPOqYL0"]
       4 |             866 | m.room.join_rules         |                                     | $Nzph2kpEaIG2FBl-1_N7Yu7Wq7q-0pLG1o-sjPOqYL0 | ["$vxSLpdP0dYGLfY4SF5ajpzm4HIGBSl8ShXFngP6dwNc"]
       3 |             865 | m.room.power_levels       |                                     | $vxSLpdP0dYGLfY4SF5ajpzm4HIGBSl8ShXFngP6dwNc | ["$RQXO-19yjf_zprhDkN4CFjj76zNyLYVwHSJRgGAarrE"]
       2 |             864 | m.room.member             | @erictroupetester:my.synapse.server | $RQXO-19yjf_zprhDkN4CFjj76zNyLYVwHSJRgGAarrE | ["$i0U5s25xbRoLLqancRLFOUcVEaLwNjZSzvvT5bU4_sw"]
       1 |             863 | m.room.create             |                                     | $i0U5s25xbRoLLqancRLFOUcVEaLwNjZSzvvT5bU4_sw | []
  (11 rows)
```



## Problem case where it fails ❌

### Reproduction


<details>
<summary>Reproduction steps</summary>

 1. Add an application service to your Synapse instance
 1. With the application service `as_token`, create a room and invite a normal user
    `POST http://localhost:8008/_matrix/client/v3/createRoom`
    ```json
    {
        "preset":       "public_chat",
        "name":         "test-synapse-12281-v0.0.1",
        "room_version": "9",
        "invite": ["@ericgittertester:my.synapse.server"],
        "initial_state": [
            {
                "type":      "m.other.state",
                "state_key": "",
                "content": {
                    "foo": "bar"
                }
            }
        ]
    }
    ```
 1. In another room, send some messages to advance the `stream_ordering` in your Synapse instance. This will add some space for us to make stream tokens in between and use with `/sync?since=abc`
 1. Join the room as the normal user
 1. As the application service, send a message in the room which we will insert our history next to
    `PUT http://localhost:8008/_matrix/client/r0/rooms/!oFSfJJEqEAtDvOnelc:my.synapse.server/send/m.room.message/123transaction`
    ```
    {
        "msgtype": "m.text",
        "body": "portal_created"
    }
    ```
 1. Use MSC2716 to insert a batch of historical messages
    `POST http://localhost:8008/_matrix/client/unstable/org.matrix.msc2716/rooms/!oFSfJJEqEAtDvOnelc:my.synapse.server/batch_send?prev_event_id=%24BhECpS4WhzG8H3QbCohaiQeBVoyK4FmBS2XYrFMkQ0c`
    ```json
    {
      "state_events_at_start": [
        {
          "content": {
            "displayname": "some-display-name-for-@maria-01234:my.synapse.server",
            "membership": "join"
          },
          "origin_server_ts": 1642055555924,
          "sender": "@maria-01234:my.synapse.server",
          "state_key": "@maria-01234:my.synapse.server",
          "type": "m.room.member"
        }
      ],
      "events": [
        {
          "content": {
            "body": "Historical 0 (batch=0)",
            "msgtype": "m.text",
            "org.matrix.msc2716.historical": true
          },
          "origin_server_ts": 1642055555924,
          "sender": "@maria-01234:my.synapse.server",
          "type": "m.room.message"
        },
        {
          "content": {
            "body": "Historical 1 (batch=0)",
            "msgtype": "m.text",
            "org.matrix.msc2716.historical": true
          },
          "origin_server_ts": 1642055555925,
          "sender": "@maria-01234:my.synapse.server",
          "type": "m.room.message"
        },
        {
          "content": {
            "body": "Historical 2 (batch=0)",
            "msgtype": "m.text",
            "org.matrix.msc2716.historical": true
          },
          "origin_server_ts": 1642055555926,
          "sender": "@maria-01234:my.synapse.server",
          "type": "m.room.message"
        },
        {
          "content": {
            "body": "Historical 3 (batch=0)",
            "msgtype": "m.text",
            "org.matrix.msc2716.historical": true
          },
          "origin_server_ts": 1642055555927,
          "sender": "@maria-01234:my.synapse.server",
          "type": "m.room.message"
        }
      ]
    }
    ```
 1. Dump the database, to easily see the gap in `stream_ordering` between the `m.room.member` `invite` and `join` events.
    ```sh
    $ psql synapse

    SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
    LEFT JOIN event_json AS ej USING (event_id)
    WHERE events.room_id ='!oFSfJJEqEAtDvOnelc:my.synapse.server' ORDER BY stream_ordering ASC;
    ```
 1. As the normal user, sync in the problem gap spot, `GET http://localhost:8008/_matrix/client/r0/sync?filter=0&timeout=30000&since=s935_205797_22_82_396_8_19_160_1`
 1. Notice that `rooms.join.<roomId>.timeline.events` and `rooms.join.<roomId>.state.events` do not include the `m.other.state` from the `initial_state` during room creation.


</details>




### Notes

Here is an overview of the events in the room:

```sh
$ psql synapse

SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
LEFT JOIN event_json AS ej USING (event_id)
WHERE events.room_id ='!oFSfJJEqEAtDvOnelc:my.synapse.server' ORDER BY stream_ordering ASC;

depth | stream_ordering |             type             |              state_key              |                   event_id                   |                   prev_events
-------+-----------------+------------------------------+-------------------------------------+----------------------------------------------+--------------------------------------------------
    11 |         -139034 | org.matrix.msc2716.insertion |                                     | $XwhXMeXCey1qng3_K2R-YhRwkYwyAo5CYnop1A-NiBA | []
    11 |         -139033 | m.room.message               |                                     | $-MW3opp5jrUNMZb82JizCmIKEHCJOImQZxJxWJiZdro | ["$XwhXMeXCey1qng3_K2R-YhRwkYwyAo5CYnop1A-NiBA"]
    11 |         -139032 | m.room.message               |                                     | $d8CvNegAiiFG8N-VlsSk-v3kI2MvCa8HZ2L7MC5WZ_8 | ["$-MW3opp5jrUNMZb82JizCmIKEHCJOImQZxJxWJiZdro"]
    11 |         -139031 | m.room.message               |                                     | $U20bBC5b0ItFzatm5eyQ2RXerkXJg7KFF7cfLvBACsw | ["$d8CvNegAiiFG8N-VlsSk-v3kI2MvCa8HZ2L7MC5WZ_8"]
    11 |         -139030 | m.room.message               |                                     | $NOozHNdRkxA5swkSa72gm0bA6ugPZc65AuE9M0IpMu0 | ["$U20bBC5b0ItFzatm5eyQ2RXerkXJg7KFF7cfLvBACsw"]
    11 |         -139029 | org.matrix.msc2716.batch     |                                     | $aDLWvwgiz_SIh4BVGs9ng1kst5d5f651GjWQ0Se5s-s | ["$NOozHNdRkxA5swkSa72gm0bA6ugPZc65AuE9M0IpMu0"]
    11 |         -139028 | org.matrix.msc2716.insertion |                                     | $4wxqppnSoQzMnm-oFmj_vfwDoRfVTkyU1xluSvN_-8I | ["$BhECpS4WhzG8H3QbCohaiQeBVoyK4FmBS2XYrFMkQ0c"]
     1 |         -139027 | m.room.member                | @maria-01234:my.synapse.server      | $2nAzxqa9gi6qQh_9z0jEZZLz6sTb4TGCo38dIHkBQm8 | []
     1 |             921 | m.room.create                |                                     | $p5PhsW7K1yP2ga4WMgsCk3vsmXMAhc4UpcUcsKOaZM0 | []
     2 |             922 | m.room.member                | @gitter-badger:my.synapse.server    | $NccDrVQCA1guredN-MUXT7Gf8m2n1_GP2t_fSMoveBU | ["$p5PhsW7K1yP2ga4WMgsCk3vsmXMAhc4UpcUcsKOaZM0"]
     3 |             923 | m.room.power_levels          |                                     | $hH_UQ3Qv8ZFHWg_Z5ge_nI5E35TKHL0JzYEP1dT6K-c | ["$NccDrVQCA1guredN-MUXT7Gf8m2n1_GP2t_fSMoveBU"]
     4 |             924 | m.room.join_rules            |                                     | $QPI5nUX19j4A09W8cszE4smeUESGZGM6KcHh8rzetCQ | ["$hH_UQ3Qv8ZFHWg_Z5ge_nI5E35TKHL0JzYEP1dT6K-c"]
     5 |             925 | m.room.history_visibility    |                                     | $lNHOvkQDv_0Fvh4smQNy6DrA6hkSfy_cEIJAHgamv80 | ["$QPI5nUX19j4A09W8cszE4smeUESGZGM6KcHh8rzetCQ"]
     6 |             926 | m.bridge                     |                                     | $3-VGN8ysxQilX1jq9KwM6mewL698Hytqwf6txGYDIVw | ["$lNHOvkQDv_0Fvh4smQNy6DrA6hkSfy_cEIJAHgamv80"]
     7 |             927 | m.other.state                |                                     | $9rgvaPPXxXfdwy1yEOaakCNiOfEpFbAm5eOq5j3GNkw | ["$3-VGN8ysxQilX1jq9KwM6mewL698Hytqwf6txGYDIVw"]
     8 |             928 | m.room.name                  |                                     | $tPjkQBhaRo3HkRruQQauSwExXXceQ0n5aDvmqWujXg4 | ["$9rgvaPPXxXfdwy1yEOaakCNiOfEpFbAm5eOq5j3GNkw"]
     9 |             929 | m.room.member                | @ericgittertester:my.synapse.server | $Rz0OJ7OWIEycom0xAe9Ky3mCWSeL1vQ4QiUy_9YpPmc | ["$tPjkQBhaRo3HkRruQQauSwExXXceQ0n5aDvmqWujXg4"]
    10 |             941 | m.room.member                | @ericgittertester:my.synapse.server | $RCs3RxSBLgmnxbpx-kf8m-Zy58_X68Y9FTvvuX3lbgI | ["$Rz0OJ7OWIEycom0xAe9Ky3mCWSeL1vQ4QiUy_9YpPmc"]
    11 |             946 | m.room.message               |                                     | $BhECpS4WhzG8H3QbCohaiQeBVoyK4FmBS2XYrFMkQ0c | ["$RCs3RxSBLgmnxbpx-kf8m-Zy58_X68Y9FTvvuX3lbgI"]
(19 rows)
```


❌ `/sync?since=s935_xxx` (anything between the invite `929` and the join `941`)

Using the `get_state_at(since_token)` sort ( `ORDER BY depth DESC, stream_ordering DESC;`), you can see that the historical messages are sorted in front of everything else and match the query.

 - The `-` red messages are filtered out because of the `WHERE 935 >= event.stream_ordering` in the SQL. 
 - `+❌` is what the SQL query selects but it's wrong
 - `+✅` is what the SQL query should have selected but it didn't

```diff
  $ psql synapse

  SELECT depth, stream_ordering, type, state_key, event_id, (ej.json::json)->'prev_events' AS prev_events FROM events
  LEFT JOIN event_json AS ej USING (event_id)
  WHERE events.room_id ='!oFSfJJEqEAtDvOnelc:my.synapse.server'
  ORDER BY depth DESC, stream_ordering DESC;
  
  depth | stream_ordering |             type             |              state_key              |                   event_id                   |                   prev_events
  -------+-----------------+------------------------------+-------------------------------------+----------------------------------------------+--------------------------------------------------
-     11 |             946 | m.room.message               |                                     | $BhECpS4WhzG8H3QbCohaiQeBVoyK4FmBS2XYrFMkQ0c | ["$RCs3RxSBLgmnxbpx-kf8m-Zy58_X68Y9FTvvuX3lbgI"]
+❌    11 |         -139028 | org.matrix.msc2716.insertion |                                     | $4wxqppnSoQzMnm-oFmj_vfwDoRfVTkyU1xluSvN_-8I | ["$BhECpS4WhzG8H3QbCohaiQeBVoyK4FmBS2XYrFMkQ0c"]
      11 |         -139029 | org.matrix.msc2716.batch     |                                     | $aDLWvwgiz_SIh4BVGs9ng1kst5d5f651GjWQ0Se5s-s | ["$NOozHNdRkxA5swkSa72gm0bA6ugPZc65AuE9M0IpMu0"]
      11 |         -139030 | m.room.message               |                                     | $NOozHNdRkxA5swkSa72gm0bA6ugPZc65AuE9M0IpMu0 | ["$U20bBC5b0ItFzatm5eyQ2RXerkXJg7KFF7cfLvBACsw"]
      11 |         -139031 | m.room.message               |                                     | $U20bBC5b0ItFzatm5eyQ2RXerkXJg7KFF7cfLvBACsw | ["$d8CvNegAiiFG8N-VlsSk-v3kI2MvCa8HZ2L7MC5WZ_8"]
      11 |         -139032 | m.room.message               |                                     | $d8CvNegAiiFG8N-VlsSk-v3kI2MvCa8HZ2L7MC5WZ_8 | ["$-MW3opp5jrUNMZb82JizCmIKEHCJOImQZxJxWJiZdro"]
      11 |         -139033 | m.room.message               |                                     | $-MW3opp5jrUNMZb82JizCmIKEHCJOImQZxJxWJiZdro | ["$XwhXMeXCey1qng3_K2R-YhRwkYwyAo5CYnop1A-NiBA"]
      11 |         -139034 | org.matrix.msc2716.insertion |                                     | $XwhXMeXCey1qng3_K2R-YhRwkYwyAo5CYnop1A-NiBA | []
-     10 |             941 | m.room.member                | @ericgittertester:my.synapse.server | $RCs3RxSBLgmnxbpx-kf8m-Zy58_X68Y9FTvvuX3lbgI | ["$Rz0OJ7OWIEycom0xAe9Ky3mCWSeL1vQ4QiUy_9YpPmc"]
+✅     9 |             929 | m.room.member                | @ericgittertester:my.synapse.server | $Rz0OJ7OWIEycom0xAe9Ky3mCWSeL1vQ4QiUy_9YpPmc | ["$tPjkQBhaRo3HkRruQQauSwExXXceQ0n5aDvmqWujXg4"]
       8 |             928 | m.room.name                  |                                     | $tPjkQBhaRo3HkRruQQauSwExXXceQ0n5aDvmqWujXg4 | ["$9rgvaPPXxXfdwy1yEOaakCNiOfEpFbAm5eOq5j3GNkw"]
       7 |             927 | m.other.state                |                                     | $9rgvaPPXxXfdwy1yEOaakCNiOfEpFbAm5eOq5j3GNkw | ["$3-VGN8ysxQilX1jq9KwM6mewL698Hytqwf6txGYDIVw"]
       6 |             926 | m.bridge                     |                                     | $3-VGN8ysxQilX1jq9KwM6mewL698Hytqwf6txGYDIVw | ["$lNHOvkQDv_0Fvh4smQNy6DrA6hkSfy_cEIJAHgamv80"]
       5 |             925 | m.room.history_visibility    |                                     | $lNHOvkQDv_0Fvh4smQNy6DrA6hkSfy_cEIJAHgamv80 | ["$QPI5nUX19j4A09W8cszE4smeUESGZGM6KcHh8rzetCQ"]
       4 |             924 | m.room.join_rules            |                                     | $QPI5nUX19j4A09W8cszE4smeUESGZGM6KcHh8rzetCQ | ["$hH_UQ3Qv8ZFHWg_Z5ge_nI5E35TKHL0JzYEP1dT6K-c"]
       3 |             923 | m.room.power_levels          |                                     | $hH_UQ3Qv8ZFHWg_Z5ge_nI5E35TKHL0JzYEP1dT6K-c | ["$NccDrVQCA1guredN-MUXT7Gf8m2n1_GP2t_fSMoveBU"]
       2 |             922 | m.room.member                | @gitter-badger:my.synapse.server    | $NccDrVQCA1guredN-MUXT7Gf8m2n1_GP2t_fSMoveBU | ["$p5PhsW7K1yP2ga4WMgsCk3vsmXMAhc4UpcUcsKOaZM0"]
       1 |             921 | m.room.create                |                                     | $p5PhsW7K1yP2ga4WMgsCk3vsmXMAhc4UpcUcsKOaZM0 | []
       1 |         -139027 | m.room.member                | @maria-01234:my.synapse.server      | $2nAzxqa9gi6qQh_9z0jEZZLz6sTb4TGCo38dIHkBQm8 | []
  (19 rows)
```




## Potential solutions

 1. The quick solution @Fizzadar came up with where we check if the new membership change matches the old membership we found, then we still consider it a new join because it is, https://github.com/Fizzadar/synapse/commit/a9a6ed945dfe02f2a583a2a4410597d83ff7c4e2
 1. Correct `get_state_at(since_token)` to actually query and sort by `stream_ordering` so it grabs the correct events in the first place
    - This work is also tracked by https://github.com/matrix-org/synapse/issues/3305 to solve a similar problem
    - This doesn't look too crazy to do. I think we just need to [adjust the SQL query](https://github.com/matrix-org/synapse/blob/287a9c1e20e1e271d69be581d6bb8ee5a8de9d88/synapse/storage/databases/main/stream.py#L1231-L1232) to first order by `stream_ordering` if don't see `from_token.topological`

**Comment by MadLittleMods** [comment]
@Fizzadar Make a PR with your fix so we can get some more feedback from maintainers. Would be good to add a test for this so it doesn't regress either. Any interest in taking it further to fix [`get_state_at(since_token)`](https://github.com/matrix-org/synapse/blob/287a9c1e20e1e271d69be581d6bb8ee5a8de9d88/synapse/handlers/sync.py#L1765)?

**Comment by Fizzadar** [comment]
@MadLittleMods great write up, thank you for pulling that all together!

I've gone ahead and change implemented the `get_state_at` change in https://github.com/matrix-org/synapse/pull/12319. I don't think the other change is worth merging as it's a workaround/hack for the real bug here, but happy to PR if there's still value in it?
